File: src/depsolver/solver/preprocessing/expand_constraints.py
from functools import reduce
from iterator_extensions import first_or_none
from model.package import Package
from model.command import Install, Uninstall
from itertools import chain

def expand_constraints_in_problem(repo_package_list,
                                  initial_state,
                                  final_state):
    repo_package_dict = repository_to_dictionary(repo_package_list)
    expanded_initial_state = expand_initial_state(repo_package_dict,
                                                  initial_state)
    expanded_final_state = expand_final_state(repo_package_dict, final_state)
    expanded_repo_package_list = expand_all_constraints(repo_package_list,
                                                        repo_package_dict)
    return (expanded_repo_package_list,
            expanded_initial_state, expanded_final_state)

def repository_to_dictionary(package_list):
    package_dictionary = {}
    for package in package_list:
        if package.name in package_dictionary:
            package_dictionary[package.name].append(package)
        else:
            package_dictionary[package.name] = [package]
    return package_dictionary

def expand_all_constraints(repo_package_list, repo_package_dict):
    def expand_package_constraints(package):
        expanded_deps = list(map(lambda disjunction: list(
                                 map(lambda constraint: list(
                                     expand_constraint(
                                         repo_package_dict, constraint)),
                                     disjunction)),
                             package.dependencies))
        expanded_conflicts = list(map(lambda conflict: list(
                                      expand_constraint(
                                          repo_package_dict, conflict)),
                                      package.conflicts))
        return Package(package.id,
                       package.name,
                       package.version,
                       package.size,
                       expanded_deps,
                       expanded_conflicts)

    return list(map(expand_package_constraints, repo_package_list))

def expand_final_state(package_dict, final_state):
    def expand_command(command):
        constraint = list(expand_constraint(package_dict, command.constraint))
        if command.is_install():
            return Install(constraint)
        else:
            return Uninstall(constraint)

    return list(map(expand_command, final_state))

def expand_initial_state(repo_package_dict, initial_state):
    def expand_specific_package(name_version_pair):
        (name, version) = name_version_pair
        packages_with_name = repo_package_dict[name]
        return next(filter_by_version(lambda v: v == version,
                                      packages_with_name)).id
    return set(map(expand_specific_package, initial_state))


def expand_constraint(repo_packages_dict, constraint):
    if constraint.package_name in repo_packages_dict:
        packages_with_name = repo_packages_dict[constraint.package_name]
        return map(lambda p: p.id,
                   filter_by_version(constraint.as_version_predicate(),
                                     packages_with_name))
    else:
        return []

def filter_by_version(version_predicate, packages_with_name):
    return filter(lambda p: version_predicate(p.version),
                  packages_with_name)

File: rust/src/depsolver/solve/preprocessing/expand_constraints.rs
use crate::depsolver::model::command::Command;
use crate::depsolver::model::constraint::PackageConstraint;
use crate::depsolver::model::package::{Package, PackageKey};
use std::collections::HashMap;
use std::collections::HashSet;
use std::iter::FromIterator;
//fn expand_constraints_in_problem(repo_package_list : Vec<Package>,
//                                 initial_state : Vec<PackageKey>,
//                                 final_state : Vec<PackageConstraint>)
//    ->  (Vec<Package>, Vec<PackageKey>, Vec<PackageConstraint>) {
//}

type NamePackageIndex<'a> = HashMap<String, Vec<&'a Package>>;

pub fn expand_constraints_in_problem(packages: Vec<Package>,
                                     initial: Vec<PackageKey>,
                                     final_state: Vec<Command>)
    -> (Vec<Package>, HashSet<i32>, Vec<Command>) {
    let index = name_to_packages_index(&packages);
    let expanded_package_list = expand_all_constraints(&packages, &index);
    let expanded_initial_state = expand_initial_state(&initial, &index);
    let expanded_final_state = expand_final_state(final_state, &index);
    (expanded_package_list,
     expanded_initial_state,
     expanded_final_state)
}

fn name_to_packages_index<'a>(package_list: &'a Vec<Package>)
    -> NamePackageIndex<'a> {
    let mut package_dictionary = NamePackageIndex::new();
    for package in package_list {
        package_dictionary.entry(package.name.clone())
                          .or_insert(vec![])
                          .push(package)
    }
    package_dictionary
}

fn expand_all_constraints(packages: &Vec<Package>, index: &NamePackageIndex) -> Vec<Package> {
    packages.into_iter()
            .map(|p| Package {
                id: p.id,
                name: p.name.clone(),
                version: p.version.clone(),
                size: p.size,
                dependencies:
                    p.dependencies.iter()
                                  .map(|dep| dep.iter()
                                                .map(|alt| expand_constraint(alt, index))
                                                .collect())
                                  .collect(),
                conflicts: p.conflicts.iter()
                                      .map(|con| expand_constraint(con, index))
                                      .collect()
            }).collect()
}

fn expand_initial_state<'a>(initial_state: &Vec<PackageKey>,
                            index: &NamePackageIndex<'a>)
    -> HashSet<i32> {
    fn expand_specific_package(index : &NamePackageIndex, package_key : &PackageKey) -> i32 {
        let PackageKey(name, version) = package_key;
        let packages_with_name = &index[name];
        packages_with_name.iter()
             .filter(|p| p.version == *version)
             .next()
             .expect("package in initial state does not exist in the repository")
             .id
    }
    return HashSet::from_iter(initial_state.iter()
                                           .map(|pk| expand_specific_package(index, pk)))
}

fn expand_final_state(final_state: Vec<Command>, index : &NamePackageIndex) -> Vec<Command> {
    fn expand_command(command: &Command, index: &NamePackageIndex) -> Command {
        match command {
            Command::Install(constraint) =>
                Command::Install(expand_constraint(&constraint, index)),
            Command::Uninstall(constraint) =>
                Command::Uninstall(expand_constraint(&constraint, index))
        }
    }

    final_state.iter()
               .map(|c| expand_command(c, index))
               .collect()
}

fn expand_constraint(constraint: &PackageConstraint, index: &NamePackageIndex)
    -> PackageConstraint {
    match &constraint {
        PackageConstraint::Unexpanded { name, version_constraint: _ } =>
            PackageConstraint::Expanded {
                possibilities: index[name].iter()
                                          .filter(|p| constraint.version_fulfils_constraint(&p.version))
                                          .map(|p| p.id)
                                          .collect()
            },
        _ => panic!("Attempted to expand constraint when already expanded.")
    }
}
